### ISSUE TITLE
Major Server Performance Regression

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#15000]
### Minetest version

```true
Minetest 5.10.0-dev (Linux)
Using LuaJIT 2.1.1703358377
Built by GCC 13.2
Running on Linux/6.1.0 x86_64
BUILD_TYPE=Release
RUN_IN_PLACE=1
USE_CURL=1
STATIC_SHAREDIR="."
```

### Irrlicht device

N/A

### Operating system and version

alpine:3.19.0 on Docker on debian 12

### CPU model

AMD EPYC 7282 16-Core Processor (VM limited to 4 cores)

### GPU model

N/A

### Active renderer

N/A

### Summary

Upgrading a server from 5.8 to 5.9 introduces massive lag, to the point where a server that was busy before becomes unusable now.

Things I have observed so far:
- Freshly restarted servers sometimes will look like they're working well for a while, but "meltdown" eventually kicks in. Sometimes MT seems to recover on its own, but sometimes this state will persist for hours.
- Before, step sizes were usually in the 0.05 to 0.10 range, with some up to 0.25 for things like the ABM tick, and rare excursions above this.  Now, step sizes are rarely smaller than 0.10, when the server is "working well", and will often exceed 1s for extended periods of time (there are very few steps smaller than 1s) when it's in "meltdown".
- Looking at Prometheus, `minetest_core_latency` is much higher (both the maximums are much higher, and the minimum baseline is higher than before) during performance "meltdowns", though `minetest_env_step_time` is indistinguishable from the way it was before.

### Steps to reproduce

"Upgrade" a busy but otherwise stable server running 5.8 to 5.9

### Further information

Getting a profiler trace of this is going to be a little tricky (I have to catch it in the act and attach to it inside the container) but I've made some progress (getting the tools installed) and am working on it.

I have also heard some anecdotal reports of similar performance regressions from others after updating to 5.9 (which suggests that this is not that specific to my environment and should be reproducible) but most of what I've heard is still vague.

**Comment by SmallJoker** [comment]
Is there any noticeable difference seen by the mod profiler?

Would you have the possibility to perform a few bisect steps to narrow down the issue?

**Comment by Warr1024** [comment]
I ran a jitprofiler capture while things were in meltdown, and didn't see anything unexpected.  I didn't compare it rigorously against a previous baseline (i.e. it's possible that _everything_ is proportionally slower, but that seemed unlikely to me) but there are no specific hotspots, similar to what I had observed before.  That suggests to me that it's not caused by Lua code, or anything that's been called by the Lua code.

**Comment by Warr1024** [comment]
Oh, sorry, missed the question about a bisect.

I'm down for a bisect, yeah, though they're pretty slow for me.  First, I probably need a way to reproduce the problem _measurably_ outside of my live public server, so it might take me a while.

**Comment by sfan5** [comment]
Getting a perf trace is likely a better use of your time than bisecting.

**Comment by Warr1024** [comment]
Haha, I actually had to gzip it (even though it's already compressed) because github "doesn't support" that file type.

[perf.data.20240818000927.gz](https://github.com/user-attachments/files/16647188/perf.data.20240818000927.gz)

This is a capture of a time period when the server was spending >70% of its time doing steps > 1 second.

**Comment by sfan5** [comment]
I'm missing the matching binaries (see [here](https://github.com/minetest/minetest/blob/master/doc/developing/misc.md#remote-profiling)).

**Comment by Emojigit** [comment]
Does this performance issue have anything to do with memory leaks? The memory consumption of my server randomly and irrecoverably spikes when it is busy, leading to my server swapping memories busily and hanging the entire server.

**Comment by Warr1024** [comment]
[debug.tar.gz](https://github.com/user-attachments/files/16650608/debug.tar.gz)

(GH says it supports .tgz, but it apparently actually doesn't)

**Comment by Warr1024** [comment]
I have not noted any significant memory consumption.  Memory usage is not very high, and is overall quite stable.

**Comment by lhofhansl** [comment]
@Warr1024 Can you easily try with the malloc_trim calls removed? (fac9aac82129edd5d77c3aca3e9e6bbf39cc3220 and 71893807b37c07bd428d71c47fd7cbd12bc82760)

**Comment by sfan5** [comment]
Thanks for the files. Unfortunately I can't spot anything that sticks out.
On the server thread, at least 21% of time is spent on ABMs, 17% on the globalstep callback and another 26% inside Lua where I have no insight.
What wouldn't be visible in the perf trace is if the code was waiting without doing any work (e.g. sleeping, locking).

Can you provide server profiler output too? (`profiler_print_interval`)
A comparison with a lag-free run would be great.

> Can you easily try with the malloc_trim calls removed?

He's on musl where this function does not exist.

> minetest_core_latency is much higher during performance "meltdowns", though minetest_env_step_time is indistinguishable from the way it was before.

this means the delay is happening inside `Server::AsyncRunStep` or `Server::Receive` but not in `ServerEnvironment::step`.
which makes even less sense since most of the juicy stuff is not in there.

**Comment by Warr1024** [comment]
I'm not seeing consistently higher CPU consumption during problems (though it's hard to tell, when MT is usually hovering around 100% no matter what) so it's actually not entirely inconsistent with IO waits.  Has 5.9 introduced any changes that would impact IO workload?

**Comment by sfan5** [comment]
I have no idea what could possibly be causing more IO or more work in general.
I scrolled through the 5.9.0 commits a bit and here are possible suspects:
* 22c4a5b2b8343c3942555c6a337e765ef84c3c4 (this is at least directly related)
* 59abf1bb426c72ab208c30b7a56b4f73d5e5cb3a (unlikely)
* 0383c44f0d6d0c18fc05640c9c018858c2ca1190 (also unlikely)

**Comment by Warr1024** [comment]
I did some significant database optimizations, and it seemed to be good again for a while.

Then, just a few minutes ago, my players reported that it had gone haywire again, and I managed to catch it in the act.  Over 75% of server steps are >= 1s, and almost none < 0.5s.

![image](https://github.com/user-attachments/assets/58fffc61-3aa6-4d77-92e0-b275195f8c3f)

Same binaries as above.

[perf.data.20240818172654.gz](https://github.com/user-attachments/files/16652376/perf.data.20240818172654.gz)

According to top, MT is at just above 100% CPU, with of spikes up to about 150%.

I'm seeing a fair amount of postgresql write activity, though the disk bandwidth isn't anywhere near peak (I know it can handle ~25M but I'm around ~0.5M).  I don't see a way to check disk busy time, but it's an SSD so seek times should be negligible.  I checked for other concurrent psql clients but there was nothing else going on.

**Comment by sfan5** [comment]
Nothing out of the ordinary again.
Time spent in "pure" LuaJIT is even higher here with >34%. Even weirder that you couldn't see anything in jitprofiler.

here's a screenshot:
![grafik](https://github.com/user-attachments/assets/8b771450-f699-476f-a583-d4691d2fae90)
for time spent in "pure" luajit (made up term; on the left) I can't tell which function caused these calls. everything else is grouped above `ServerThread::run()`

**Comment by Warr1024** [comment]
34% Lua time wouldn't account for step time going up 400% to 2000%.

Sounds like it's not caused by added compute workload, but some kind of problem with a wait.

**Comment by Desour** [comment]
>Time spent in "pure" LuaJIT is even higher here with >34%. 

FYI, this is usually the garbage collector.
(If you aggregate by thread, you can see that it's in the server thread, but not under the thread's Thread::threadProc symbol. That's probably because the gc steps run in a signal handler.)

Some perf data for when there's no long server steps would be useful for comparison.

It might be worth noting that longer steps do not necessarily mean that there's more work to do. It can also be that the work is just not spread as well across multiple steps. This can likely not be seen in a sampling profiler.
And if you use dtime (which is the time the last step took, AFAIK) to determine how much work you do in the current step, you can easily make lag spikes persist longer, because long steps result in more work in the next step, making that one longer again. Idk how abms perform in this regard.

**Comment by Warr1024** [comment]
> And if you use dtime (...) to determine how much work you do in the current step, you can easily make lag spikes persist longer

Yeah, I've done a decent job of avoiding that, at least.

> Idk how abms perform in this regard.

ABMs cause a single step to take +0.2s once per second, by default configuration.  Since the timers never desync, you'll always have a thundering herd once per second even with non-default config, unless ABMs let you do some crazy stuff like setting interval to like 0.997s or something.

It's not much of a problem, though; it's mostly accounted for.  I should have 75% of my time spent in 0.05s steps, and 25% doing that one 0.25s step per second.

A healthy lagometer for a default-configured ABM-heavy game looks like:

![image](https://github.com/user-attachments/assets/ffec7441-b74c-4e1d-ab40-0cf12d38b89a)

Most of the time is spent doing the finest possible steps, and the ABM lag is a small predictable hump above that.  There should be almost no excursions to 0.5 or higher.

The lagometer mod's top bucket is the 1s one, so if you get a 5s step, it just adds 5s to the 1s bucket.  So anything in 1s+ is bad, but it doesn't tell you exactly how bad.  I assumed that at that point, it doesn't matter that much.

> FYI, this is usually the garbage collector.

Is it worth going far out of the way to write GC-reducing code, like code that tries to reuse data structures instead of discarding and rebuilding them?  I've done this _lightly_ but considered it a premature optimization so I didn't go far down that route.

**Comment by sfan5** [comment]
> (If you aggregate by thread, you can see that it's in the server thread, but not under the thread's Thread::threadProc symbol. That's probably because the gc steps run in a signal handler.)

The screenshot is filtered down to only the server thread.
And I think the cause is that JIT-generated code doesn't have debug information so the stack trace can be reconstructed.
If @Warr1024 were to compile with Lua this won't happen IIRC.

**Comment by Warr1024** [comment]
I did a bunch of database maintenance stuff, like vacuuming and such, and it improved things to some extent.  When the server is working well, it's working quite well (possibly better than 5.8 did), it's just very easy to get into a bad state.  Things like intensive operations in the background can cause problems.

I tried reverting back to 5.8, just to make sure that the server maintenance changes weren't a confound.  It's much better in 5.8, so it really is a difference between 5.8 and 5.9, whether the server is otherwise tuned well or not.

The biggest difference seems to be when disk-intensive background operations are happening, like backups (entirely external to MT).  Under 5.9, this was one of the biggest performance killers, but under 5.8, MT isn't bothered by them at all.

All of this points to increased disk/database IO being the big difference.

**Comment by sfan5** [comment]
To be clear without more information I don't think we'll be able to determine the cause.
So bisecting is the next best option.

**Comment by Warr1024** [comment]
It looks like I've got the issue mitigated enough on my server that it's effectively not a problem there anymore.  The fact that there's a performance difference between 5.8 and 5.9 is still very suggestive that there's a problematic underlying change in 5.9, but I never found a smoking gun.

I attempted a bisect, but I had never been able to reproduce the problem in the lab, and by the time I tried to bisect, I already had enough mitigations in production that I couldn't repro it there either.

# Background

My server runs on a single Contabo VPS S SSD instance.  It's a 4-core VM with 8GB of RAM and a 200GB SSD.  The performance of all these components is pretty good, and I haven't noticed anything like cross-VM resource sharing issues.

Everything runs inside containers on a single VM, including the MT server, databases, backups and all other scripts.  The MT server uses postgresql for everything: auth, players, map, mod_storage.

My backup process runs while the server remains online (one of the key benefits of postgresql) every couple hours, continuously throughout the day.  To backup the world, I migrate it from postgresql to sqlite3, which produces a playable/testable snapshot of the world.

# Mitigations and Corrections

## server_unload_unused_data_timeout

The default value for this is 29.  I had been using 600 before, but recently removed a bunch of old settings that were based on outdated performance advice.  This one, it turns out, was still very relevant.  Raising it back to 600 wasn't enough to cover the 5.8/5.9 performance difference but it made both faster, especially while plyaers were NOT moving around.

We should probably consider changing the default, especially for dedicated servers (or make a different default kick in for dedicated servers).

## Garbage Players in Database

I found a lot of players in the database with no corresponding auth.  This appears to be caused by trying to remove_player and remove_player_auth at the same time, immediately after a player has left the server.  Player auth was getting removed but the player was not, possibly because players being written back to the database may be delayed.  I added a delayed retry of remove_player for a few seconds, and haven't seen new garbage players since.

Because I was using minetestserver --migrate-players as part of the background process, removing thousands of garbage players sped up this step of the process a lot.

## Garbage mod_storage Entries

I have several mods that store per-player data "inside out" in mod_storage, so that it's accessible when players are not online, or so that it's tied to the player's auth and not necessarily their player character (I want it cleaned out only on remove_player_auth, not remove_player).

I noticed that a lot of old players that were being automatically pruned (fly-bys that didn't complete true registration, or old players pruned for inactivity) were still filling a few mods with thousands of entries.  I added garbage collection for these mods.

The total number of entries was reduced by about 8%, though the total disk space usage was only reduced by about 1%.  Nevertheless, it contributed to faster backups.  It seems that minetestserver --migrate* may be executing a lot of individual per-row statements with implicit transactions, and lots of small rows written this way would cause disproportionaly high IO.

## Backup Optimizations

I made a few other optimizations to the backup process, with the end effect being a roughly 10x speedup, along with associated reduction in background IO workload from it.

I had already had switched from using minetestserver --migrate for map.sqlite for backups long ago to use custom scripts and direct sql access, in order to make backups dedupe better, but had noticed a performance improvement from it, as well as writing everything in a single transaction.  I made further improvements by replacing "update, insert or ignore" with "insert on conflict" statements, batching 100 rows per statement.

For now, I'm continuing to use minetestserver --migrate* for the other 3 databases, but I got further speedup by pre-initializing the databases with schema and setting journal_mode to WAL before migrating into them.

I also ran a one-time vacuum of the map database, which had minor impact at the time, though I want to avoid doing this routeinely because it worsens deduping.

## Postgresql Tuning

Apparently the postgresql docker containers ship with postgresql RAM cache configured to 128MB.  Their documentation recommends 15% to 25% of total RAM, or 2GB on my server.  I bumped up the limit, since I had plenty of RAM to spare.  I saw only minor performance improvements from it, though.  In theory, over a long enough runtime (or across MT server restarts without restarting postgresql) I'm hoping to see a little improvement.

I messed around with vacuuming a bit, but it turns out postgresql was already auto-vacuuming out of the box and the default settings are probably already fine.

# Conclusions

There's still something different between 5.8 and 5.9.  Ironically it seems like 5.9 is a little more CPU-efficient, but somehow it's either a bit less IO-efficient, or a bit more UI-bound and more subject to outside interference.  I haven't been able to reproduce the problem in the lab, though, nor can I capture database activity in production.  It seemed like in the past, if IO was slow, MT would load terrain more slowly, but for the most part the main thread could still respond.  Now, any IO lag is felt right in the main thread.

Backend migration has more use-cases than just one-time migrations, and we should probably make it more efficient so it can be useful in other use-cases like making backups, snapshots for diagnostics, etc.  In particular the way it's doing sqlite writes seems to be very naive, which probably could stand to be improved considering how sqlite is the "main" database choice.

Also, considering sqlite's status as a "main" backend choice, it may make sense to have an option to enable WAL mode.  Currently it requires creating the database manually in an external process, which is awkward and fragile.

The fact that players can get written to the database after they've left the server and I call remove_player seems like a bug.

**Comment by Warr1024** [comment]
After a few days of quiet, I caught another meltdown.  It seems like the meltdown condition is separate from the overall IO-bound issue.  I also watched iotop for a while and there was very little disk activity ... it even entered a more or less completely quiescent state, while server performance was still in the shitter.

Memory usage does seem higher than normal.  The entire map sqlite would fit in memory at that size, which doesn't seem right for the 4 players I had on around that time.

![screenshot_20240827_173148](https://github.com/user-attachments/assets/4e7f4634-2424-489e-95ce-0da729b5c6b7)
![20240827173212](https://github.com/user-attachments/assets/e5028068-89e3-4d9c-b9c2-b6253ce4cdc0)

**Comment by SmallJoker** [comment]
> Before, step sizes were usually in the 0.05 to 0.10 range, with some up to 0.25 for things like the ABM tick, and rare excursions above this. Now, step sizes are rarely smaller than 0.10, when the server is "working well", 

Random idea, solely based on that you mentioned RAM. Does the problem vanish if you insert an early return to `void TriggerMemoryTrim()`, preventing calling `malloc_trim`?

**Comment by Desour** [comment]
FWIW, you can also try a tracing profiler, e.g. tracy (<https://github.com/wolfpld/tracy>).
Start with this https://github.com/Desour/minetest/tree/tracy_profiling (and rebase), and add zonescopes to narrow the issue down. Lua is also supported by tracy.

**Comment by Desour** [comment]
>(and rebase)

It's rebased now, FYI. (It was non-trivial.) :)

**Comment by Warr1024** [comment]
Well, after tracking down a lot of little problems that had a small impact, I did find one huge issue.

![image](https://github.com/user-attachments/assets/3d741a0b-2ee5-4774-a7b1-945adab93516)

That one long orange "C" part there is in the vector constructor for when the engine is pushing a vector into MT.  It seems like vector construction is very slow now.  It looks like it's happening inside the callback for `register_on_liquid_transformed`; we process up to 100,000 liquid transformation nodes per cycle according to the default conf, and those all get pushed into Lua.  At some point, the constructor for vectors seems to have gotten a lot more expensive (i.e. adding the metatable).  Making large optimizations on what liquid transformed positions were used for had very little impact.

Should I spin that off as a separate issue, maybe?

**Comment by sfan5** [comment]
The last change to that was b38ffdec27, before 5.8.0. I bet benchmarks would show that the new way is indeed faster.

> It looks like it's happening inside the callback for `register_on_liquid_transformed`

That callback is extraordinarily inefficient just by existing really....

> [...] and those all get pushed into Lua

That is, if you have at least one registered callback. is that the case?

**Comment by Warr1024** [comment]
> if you have at least one registered callback. is that the case?

Indeed I do.  In fact it was for a mechanic of this game that I needed the API to be added in the first place.

> extraordinarily inefficient just by existing
It's very inefficient, almost as bad as the one-callback-per-pos of ABM/LBMs.  On the other hand, I'm uncertain how the callback could be made more efficient.

In many cases, I'm hashing those positions to look up in a LUT, so if I got them as an array or set of integers, it might actually work out a lot better.  I might have to make some breaking changes to a game-layer API, but from what I've seen, that API is barely used outside the game itself, so if I need to make breaking changes, now is better than later.

**Comment by Warr1024** [comment]
Confirmed that it's NOT from vectors being passed to me via register_on_liquid_transformed: I still get that huge vector constructor thing in my flame graph even when that callback is getting zero-length lists.

**Comment by Warr1024** [comment]
Well, I managed to find a major source of lag.  The "total meltdown" state was apparently caused by game code in the green-outlined section.

![image](https://github.com/user-attachments/assets/f37df8c3-177f-4d86-a825-a2d45b147b31)

I was able to fix this (affected 5.8 and 5.9 under specific circumstances).

The fact that it shows up as like 10% of the total flame graph even when it's responsible for more than half of the lag on the server makes this kind of thing really hard to debug; the above graph was captured *during* the meltdown.  I basically started auditing code everywhere that showed up on the graph at all, and was a bit lucky about what I ran across.  It would be nice to have better tools.  I'm still especially curious where all those vector.lua calls from the C side are coming from.

Overall, things are more tolerable, but something still isn't quite right.  It seems like Minetest takes longer to _load_ areas of the map overall.  Each time I find a problem in game or mod code, I should be able to get a clearer picture of how 5.9 performs compared to past versions, without all the random sources of lag confounding my results.

I suppose I'll just keep poking at this until either I find a smoking gun, go a few weeks without any progress, or somebody points me at a better tool.  Tracey sounded a bit too involved at the time, but I might give it a try in a few days if things are still problematic.

**Comment by Warr1024** [comment]
I originally spun off #15125 as a separate issue because this issue is about new performance issues in 5.9, whereas that one was about a performance issue that affects all versions.  However, it seems like they are related.

I've been back and forth between 5.8 and 5.9 several times throughout this, and what I'm seeing now looks like:

- When the _database_ is busy (e.g. a pg_dump is running), both 5.8 and 5.9 have serious lag (**15125**).
- When the database is **not** busy but the _disk_ is (e.g. a background job is running, but pgsql is quiescent and has a well-filled cache), only 5.9 experiences sporadic lag spikes, while 5.8 runs smoothly (**NOT** part of **15125**).

This latter observation could be consistent with disk access that's not one of mod_storage, players, auth, or map, such as one of the text files in the world.  If e.g. we were waiting on a flush-to-disk while writing env_meta.txt while holding env lock in 5.9, but that flush wasn't happening, or was happening off the main thread or outside a lock in 5.8, that would be enough to explain the difference in observations between what I'm seeing here and what's already covered in #15125.

**Comment by Warr1024** [comment]
I no longer believe this is a regression specific to 5.9.  Everything I've been seeing over the past couple of weeks suggests that the problem is adequately explained by a combination of:

- #15125
- exogenous factors like VPS performance degrading (host system may be busier than before)

In particular I haven't been able to demonstrate measurable performance problems "in the lab" that wouldn't be explained by these.
